### ISSUE TITLE
Add ability to hide code cell input in CoCalc Jupyter Notebook

### ISSUE THREAD

**Issue by profbrady** [sagemathinc/cocalc#3835]
When using CoCalc for teaching programming it is sometimes helpful to be able to hide individual input cells while still showing the output. This would be nice specifically when using the new widgets functionality.

**Comment by williamstein** [comment]
Can you check to see if this is implemented in any way using jupyterlab or jupyter classic, or an extension thereof?  I ask, because I just want to make sure our implementation is compatible, so you can hide with jupyterlab (say) and it'll be hidden when you open the same ipynb in cocalc, and conversely.  This just involves making sure the underlying metadata entry in the cell is set the same way.

**Comment by profbrady** [comment]
The “Hide Cells” extension works. The cell metadata reads the following after adding the extension and hiding a cell:

{
 "hide_input": true
}

If the CoCalc notebook would allow this cell metadata to hide input I would be happy. Maybe add it to the Edit menu like toggling view and write protection were?

Brian

**Comment by williamstein** [comment]
Related note -- the top rated answer [here](https://stackoverflow.com/questions/31517194/how-to-hide-one-specific-cell-input-or-output-in-ipython-notebook) talks about nbconvert supporting some tags, which will hide an entire cell (not just input), in a way that isn't compatible with the ["Hide Cells" extension](https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/hide_input/readme.html), evidently.  Gees!!

I found the nbconvert tag docs, which seem to exist **only in the changelog** (!), and they specify how to hide input/output/complete cells, and it's totally different than the "Hide Cells" extension:

https://nbconvert.readthedocs.io/en/latest/changelog.html?highlight=remove_cell#tag-based-element-filtering

How important is it to support the Hide Cell unofficial extension?  I could:

1. Support the official nbconvert tags, or
2. Support the "Hide Cells" extension approach (but not tags), or
3. Just do both (be redundant)...   which will lead its own confusion at some point, and necessitate making arbitrary choices (e.g., what if the data in the tags and in metadata with Hide Cells is there in the ipynb and they conflict?).

Any thoughts?

**Comment by profbrady** [comment]
It is not important to me that the unofficial “Hide Cells” extension is supported. I think support for the official nbconvert tags should work. What is important to me is that I can hide input cells from students while still showing the output. 

Brian

**Comment by williamstein** [comment]
Cool, I'll do just option 1.

Also, I could add a keyboard short.  If so, any suggestions...?

**Comment by williamstein** [comment]
I just implemented the part that toggles the tags...

![Screenshot 2019-05-20 at 9 16 34 PM](https://user-images.githubusercontent.com/1276278/58068079-bcf9f080-7b44-11e9-8d16-7680a2338c48.png)

I realize the word "selected" is inconsistent here.  I'll fix that.

**Comment by jasongrout** [comment]
There is a standard for metadata for hiding cell inputs, which we've implemented in JupyterLab and I believe nteract: https://nbformat.readthedocs.io/en/latest/format_description.html#cell-metadata

If a cell has metadata `{'jupyter': {'source_hidden': true}}`, its input should be hidden. In JupyterLab, we interpret that as collapsing the cell input (i.e., the user can expand the cell input to see the input if they want, it's not that the cell input is completely inaccessible).

**Comment by jasongrout** [comment]
FYI, the classic notebook does *not* respect this `source_hidden` field. In the notebook format, all metadata fields are optional ("All metadata fields are optional. While the types and values of some metadata fields are defined, no metadata fields are required to be defined. Any metadata field may also be ignored." from the top of https://nbformat.readthedocs.io/en/latest/format_description.html#the-notebook-file-format)

**Comment by jasongrout** [comment]
Here is how the metadata corresponds to the visual state in JLab. The metadata is shown on the left, the visual state is changed on the right.
![cellhiding](https://user-images.githubusercontent.com/192614/58068519-2e579480-7b81-11e9-928e-8891690a03ab.gif)

**Comment by williamstein** [comment]
This seems incompatible with nbconvert's use of tags?  Is it?

**Comment by williamstein** [comment]
It's not really hidden - just collapsed.  Hmm.  Brian do you want hidden or
collapsed?  I recall things like genuinely fully hidden (%auto) cells being
super important to sagenb users and want to at least support that to...

Thanks for looking into this!!

**Comment by jasongrout** [comment]
nbconvert doesn't specify what tags correspond to hiding cells/inputs/outputs, right? They just make it possible to flexibly say at runtime which subsets of cells should be hidden, have their inputs hidden, etc. At least that's how I understand it (I haven't used it before).

We do have a cell-tags extension that hopes to make working with subsets of cells much easier (e.g., selecting and operating on groups of cells based on tag search expressions). That would make things more like nbconvert, in that you'd be able to easily select all cells with a given tag, then hide all of their inputs together.

**Comment by profbrady** [comment]
I think I’m looking for collapsed. 

Brian

**Comment by williamstein** [comment]
> nbconvert doesn't specify what tags correspond to hiding cells/inputs/outputs, right?

OK, thanks for the clarification.  I found  [the docs](https://nbconvert.readthedocs.io/en/latest/changelog.html?highlight=remove_cell#tag-based-element-filtering) very hard to follow.  I'll just ignore nbconvert and this unofficial extension and for now plan to implement something exactly like JupyterLab (since that also satisfies @profbrady).

**Comment by williamstein** [comment]
@jasongrout -- one other situation where a choice must be made is what happens when you evaluate a cell whose input or output is hidden.  I think this is what jupyterlab does:

- the input stays hidden
- the output is unhidden.

**Comment by williamstein** [comment]
- if it is a markdown cell then the input is revealed on evaluation.

**Comment by williamstein** [comment]
@jasongrout -- some further comments.

1. For JupyterLab what is the relationship between metadata.collapsed = true and metadata.jupyter.outputs_hidden = true?

2. JupyterLab's *View* menu has fully 9 lines (including a divider) dedicated to changing these outputs_hidden and input_hidden fields:

![image](https://user-images.githubusercontent.com/1276278/58132497-49a0bf00-7bd6-11e9-9042-44d3ce5fbc17.png)

One might even argue that it's wrong to put these entries in the "View" menu, since they impact the state of the underlying document, whereas everything else in the View menu impacts how you're looking at it (e.g., showing a toolbar or not).   Also, maybe 9 lines makes the menu longer and more overwhelming?  So I'm not implementing this, and instead doing the following (which I don't love either):

![image](https://user-images.githubusercontent.com/1276278/58132570-79e85d80-7bd6-11e9-805e-8d6926076890.png)

**Comment by williamstein** [comment]
In any case, something that is at least enough for @profbrady will be in the next release.

**Comment by williamstein** [comment]
This is now live on cocalc.com
